source/common/listener_manager/listener_impl.cc

```diff
@@ -487,6 +501,110 @@ ListenerImpl::ListenerImpl(ListenerImpl& origin,
   }
 }
 
+ListenerImpl::ListenerImpl(ListenerImpl& origin, const FilterChainRefVector& added_filter_chains,
+                           const absl::flat_hash_set<absl::string_view>& removed_filter_chains,
+                           absl::optional<absl::string_view>& fcds_version_info,
+                           bool workers_started, absl::Status& creation_status)
+    : parent_(origin.parent_), addresses_(origin.addresses_), socket_type_(origin.socket_type_),
+      bind_to_port_(origin.bind_to_port_), mptcp_enabled_(origin.mptcp_enabled_),
+      hand_off_restored_destination_connections_(origin.hand_off_restored_destination_connections_),
+      per_connection_buffer_limit_bytes_(origin.per_connection_buffer_limit_bytes_),
+      listener_tag_(origin.listener_tag_), name_(origin.name_),
+      added_via_api_(origin.added_via_api_), workers_started_(workers_started),
+      maybe_stale_hash_(origin.maybe_stale_hash_), tcp_backlog_size_(origin.tcp_backlog_size_),
+      max_connections_to_accept_per_socket_event_(
+          origin.max_connections_to_accept_per_socket_event_),
+      validation_visitor_(
+          added_via_api_ ? parent_.server_.messageValidationContext().dynamicValidationVisitor()
+                         : parent_.server_.messageValidationContext().staticValidationVisitor()),
+      ignore_global_conn_limit_(origin.ignore_global_conn_limit_),
+      bypass_overload_manager_(origin.bypass_overload_manager_),
+      start_listener_without_warming_(origin.start_listener_without_warming_),
+      created_by_fcds_(true),
+      // Used in case where FCDS is received while server is initializing and the listener has a
+      // dependency on FCDS update to arrive to complete initialization.
+      listener_init_target_(std::make_shared<Init::TargetImpl>(
```

Comment:
This is no different than a regular filter initialization I think? If so, I'd suggest reusing the code - maybe with a shared private constructor (there's a lot of fields, so not sure how easy that is).